swarm/src/lib.rs

```diff
@@ -676,14 +691,16 @@ where
                     }
                     let peer_id = connected.peer_id;
                     let endpoint = connected.endpoint;
-                    this.behaviour.inject_connection_closed(
-                        &peer_id,
-                        &id,
-                        &endpoint,
-                        handler.into_protocols_handler(),
-                    );
-                    if num_established == 0 {
-                        this.behaviour.inject_disconnected(&peer_id);
+                    if !this.banned_connections.remove(&id) {
+                        this.behaviour.inject_connection_closed(
+                            &peer_id,
+                            &id,
+                            &endpoint,
+                            handler.into_protocols_handler(),
+                        );
+                        if num_established == 0 {
```

Comment:
What if there is another connection to the same peer, but that connection is in `banned_connections`? `num_established` would be `!= 0` and thus `inject_disconnected` would never be called. Correct?

If that is the case, we could carry the concrete `ConnectionId`s in `ConnectionClosed::num_established`. In case `num_established != 0`, we also check whether the remaining `ConnectionId`s are in `banned_connections` and if so, call `inject_disconnected` as well.

Reply:
That's a very good catch. I'll think how to solve it tomorrow and hopefully have this ready by your monday

Reply:
I’m sure you’re already planning on doing this, so please forgive the note if so.

We should double check all other usages of `num_established` – particularly the [`inject_connected`](https://github.com/libp2p/rust-libp2p/blob/eefc25d5dffdb464afa5d30e94d7d0eaad42b73e/swarm/src/lib.rs#L669) call. Maybe we can even add a helper method to make this less tricky? something like `num_established_unbanned_conns` (naming?).

There’s some ambiguity whether network events should use the raw `num_established` or they should use the `num_established_unbanned_conns` form instead. I’m not sure. My guess is it should use the `num_established` because that’s the real number of established connections. And otherwise you may get weird cases where the event says 0 established, but some `is_connected` method returns true (maybe I’m wrong here).

Reply:
My politic here would be: do the changes that make us fulfill the contract since the behaviour's perspective. From the swarm's API, to ignore banned peers we would need to rework way more than what I think is in scope of the current PR. For completeness:
- `network_info()` with all it's counters and etc will now and then not match the behaviour's view.
- `num_established` will also differ when banned connections are still open.
- `is_connected` might return `true` for a banned peer.
However unexpected, none of these break a contract. The info about the network is untainted with the banning info. From my perspective that's fine.